bigtable/beam/keyviz-art/src/main/java/keyviz/ReadData.java

```diff
@@ -0,0 +1,264 @@
+import org.apache.hadoop.hbase.filter.MultiRowRangeFilter.RowRange;
+import org.apache.hadoop.hbase.util.Bytes;
+import org.joda.time.Duration;
+
+public class ReadData {
+
+  static final long KEY_VIZ_WINDOW_MINUTES = 15;
+  static final long ONE_MB = 1000 * 1000;
+  static final long ONE_GB = 1000 * ONE_MB;
+  static final String COLUMN_FAMILY = "cf";
+
+  static final long START_TIME = getStartTime();
+
+  public static void main(String[] args) {
+    ReadDataOptions options =
+        PipelineOptionsFactory.fromArgs(args).withValidation().as(ReadDataOptions.class);
+    Pipeline p = Pipeline.create(options);
+    CloudBigtableTableConfiguration bigtableTableConfig =
+        new CloudBigtableTableConfiguration.Builder()
+            .withProjectId(options.getBigtableProjectId())
+            .withInstanceId(options.getBigtableInstanceId())
+            .withTableId(options.getBigtableTableId())
+            .build();
+
+    // Initiates a new pipeline every second
+    p.apply(GenerateSequence.from(0).withRate(1, new Duration(1000)))
+        .apply(ParDo.of(new ReadFromTableFn(bigtableTableConfig, options)));
+    p.run().waitUntilFinish();
+  }
+
+  public static class ReadFromTableFn extends AbstractCloudBigtableTableDoFn<Long, Void> {
+
+    List<List<Float>> imageData = new ArrayList<>();
+    String[] keys;
+
+    public ReadFromTableFn(CloudBigtableConfiguration config, ReadDataOptions readDataOptions) {
+      super(config);
+      keys = new String[Math.toIntExact(getNumRows(readDataOptions))];
+      downloadImageData(readDataOptions.getFilePath());
+      generateRowkeys(getNumRows(readDataOptions));
+
+    }
+
+    @ProcessElement
+    public void processElement(PipelineOptions po) {
+      // Determine which column will be drawn based on runtime of job.
+      long timestampDiff = System.currentTimeMillis() - START_TIME;
+      long minutes = (timestampDiff / 1000) / 60;
+      int timeOffsetIndex = Math.toIntExact(minutes / KEY_VIZ_WINDOW_MINUTES);
+
+      ReadDataOptions options = po.as(ReadDataOptions.class);
+      long count = 0;
+
+      List<RowRange> ranges = getRangesForTimeIndex(timeOffsetIndex, getNumRows(options));
+      if (ranges.size() == 0) {
+        return;
+      }
+
+      try {
+        // Scan with a filter that will only return the first key from each row. This filter is used
+        // to more efficiently perform row count operations.
+        Filter rangeFilters = new MultiRowRangeFilter(ranges);
+        FilterList firstKeyFilterWithRanges = new FilterList(new FirstKeyOnlyFilter(),
+            rangeFilters);
+        Scan scan =
+            new Scan()
+                .addFamily(Bytes.toBytes(COLUMN_FAMILY))
+                .setFilter(firstKeyFilterWithRanges);
+
+        Table table = getConnection().getTable(TableName.valueOf(options.getBigtableTableId()));
+        ResultScanner imageData = table.getScanner(scan);
+
+        for (Result row : imageData) {
+          count++;
+        }
```

Comment:
Isn't there a better way to do this?  If not, perhaps an explanation of what's going on would be useful?

Reply:
Yeah, this is the way to do it. Can leave a comment